src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -632,104 +518,39 @@ private ODataBatchOperationRequestMessage CreateOperationRequestMessageImplement
-        private ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId)
+        private void UpdateAfterWriteStartChangeset()
         {
-            this.WritePendingMessageData(true);
-
-            // In responses we don't need to use our batch URL resolver, since there are no cross referencing URLs
-            // so use the URL resolver from the batch message instead.
-            this.CurrentOperationResponseMessage = ODataBatchOperationResponseMessage.CreateWriteMessage(
-                this.rawOutputContext.OutputStream,
-                /*operationListener*/ this,
-                this.payloadUriConverter.BatchMessagePayloadUriConverter,
-                this.container);
-            this.SetState(BatchWriterState.OperationCreated);
-
-            Debug.Assert(this.currentOperationContentId == null, "The Content-ID header is only supported in request messages.");
-
-            // write the operation's start boundary string
-            this.WriteStartBoundaryForOperation();
-
-            // write the headers and request separator line
-            ODataBatchWriterUtils.WriteResponsePreamble(this.rawOutputContext.TextWriter, changeSetBoundary != null, contentId);
-
-            return this.CurrentOperationResponseMessage;
-        }
-
-        /// <summary>
-        /// Writes all the pending headers and prepares the writer to write a content of the operation.
-        /// </summary>
-        private void StartBatchOperationContent()
-        {
-            Debug.Assert(this.CurrentOperationMessage != null, "Expected non-null operation message!");
-            Debug.Assert(this.rawOutputContext.TextWriter != null, "Must have a batch writer!");
-
-            // write the pending headers (if any)
-            this.WritePendingMessageData(false);
-
-            // flush the text writer to make sure all buffers of the text writer
-            // are flushed to the underlying async stream
-            this.rawOutputContext.TextWriter.Flush();
+            // reset the size of the current changeset and increase the size of the batch.
+            this.ResetChangeSetSize();
+            this.InterceptException(this.IncreaseBatchSize);
+            this.isInsideSubBatch = true;
         }
 
         /// <summary>
-        /// Disposes the batch writer and set the 'OperationStreamRequested' batch writer state;
-        /// called after the flush operation(s) have completed.
+        /// Perform updates after changeset is ended.
         /// </summary>
-        private void DisposeBatchWriterAndSetContentStreamRequestedState()
+        private void UpdateAfterWriteEndChangeset()
         {
-            this.rawOutputContext.CloseWriter();
-
-            this.SetState(BatchWriterState.OperationStreamRequested);
+            // Reset the cache of content IDs here. As per spec, content IDs are only unique inside a change set.
+            this.payloadUriConverter.Reset();

```

Comment:
As in reading, this should only be reset for this.outputContext.MessageWriterSettings.Version <= ODataVersion.V4

Reply:
Done. Also, to align with reading, moved this Reset() to UpdateAfterWrteStartChangeset, which is part of the start changeset processing.

---
In reply to: [145852007](https://github.com/OData/odata.net/pull/960#discussion_r145852007) [](ancestors = 145852007)